### PR TITLE
Fix #1961: Honor call-site @inline / @noinline

### LINKED ISSUE #1961: Honor call-site @inline and @noinline
The feature was introduced in Scala/JVM in https://github.com/scala/scala/pull/4801

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

**Comment by gzm0** [comment]
@sjrd, I was conflicted on whether to bump the IR version here or not: IIUC this is a fully forward compatible change (both binary format and spec): The additional bits will simply be ignored by both the deserializer and the optimizer. And ignoring the inline hints is acceptable, specd behavior.

File: compiler/src/test/scala/org/scalajs/nscplugin/test/CallSiteInlineTest.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+class CallSiteInlineTest extends JSASTTest {
+  @Test
+  def testInline: Unit = {
+    val flags = {
+      """
+      object A {
+        println("F"): @inline
+      }
+      """.extractOne("println call") {
+        case js.Apply(flags, _, js.MethodIdent(name), _) if name.simpleName.nameString == "println" =>
```
Comment (gzm0) [inline compiler/src/test/scala/org/scalajs/nscplugin/test/CallSiteInlineTest.scala:32]: We should arguably construct the simple name beforehand here. However, I ran into some module initialization issues inside the `Name` object. Rings a bell?

**Comment by sjrd** [comment]
> @sjrd, I was conflicted on whether to bump the IR version here or not: IIUC this is a fully forward compatible change (both binary format and spec): The additional bits will simply be ignored by both the deserializer and the optimizer. And ignoring the inline hints is acceptable, specd behavior.

I think it's fine not to bump. Then again, bumping would allow to add flags to `New`, which would address your comment at https://github.com/scala-js/scala-js/issues/1961#issuecomment-1685241828

File: compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala [hunk 2]
```diff
@@ -3281,35 +3281,53 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
           genApplyStatic(sym, args.map(genExpr))
```
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:3281]: This one probably wants the hints as well, no?

File: linker/shared/src/test/scala/org/scalajs/linker/OptimizerTest.scala [hunk 3]
```diff
@@ -561,6 +561,67 @@ class OptimizerTest {
       assertFalse(findClass(moduleSet, "Witness").isDefined)
     }
   }
+
+  /** Returns whether the method got inlined. */
```
Comment (sjrd) [inline linker/shared/src/test/scala/org/scalajs/linker/OptimizerTest.scala:565]: I think this comment is lying ;)

File: compiler/src/test/scala/org/scalajs/nscplugin/test/CallSiteInlineTest.scala (hunk 1 above)
Comment (sjrd) [reply compiler/src/test/scala/org/scalajs/nscplugin/test/CallSiteInlineTest.scala:32]: Hum. That doesn't ring a bell, no. :s

**Comment by gzm0** [comment]
> One thing we probably want to do as a follow-up is to mark call-site @inline in GenExports for methods of JS classes when they have no overload. 

Shouldn't we only do this if the export is not overridden with more overloads in a subclass (IIUC not a condition we can check in the compiler)? Otherwise we might duplicate the method body.

**Comment by sjrd** [comment]
When that happens, I'm pretty sure we emit a JS-level `super` call, so the body won't be duplicated. (The JS `super` call cannot be inlined since it is at the JS level.)

**Comment by gzm0** [comment]
> When that happens, I'm pretty sure we emit a JS-level super call, so the body won't be duplicated. (The JS super call cannot be inlined since it is at the JS level.)

Right. I was thinking about JSExports in Scala classes. But indeed in JS classes in that case we'll call super.

**Review by sjrd — APPROVED** [review]
One Scalastyle issue, but otherwise LGTM.

**Pull request merged**